--- origin/tinybendygrad/schedule/prepare.bend	2026-10-04 22:19:06
+++ working/tinybendygrad/schedule/prepare.bend	2026-10-04 22:19:06
@@ -312,7 +312,7 @@
   O.PMEntrys{[
     O.PMEntry{0, [O.OpsAFTER{}], [O.OpsSTORE{}]},  # 0 AFTER rej=STORE slen=1 rlen=2 any=0 nalts=1 name=after
     O.PMEntry{1, [O.OpsSTAGE{}], Nil{}},  # 1 STAGE rej=- slen=1 rlen=1 any=0 nalts=1 name=after
-    O.PMEntry{2, [O.OpsEXP2{}, O.OpsLOG2{}, O.OpsSIN{}, O.OpsSQRT{}, O.OpsRECIPROCAL{}, O.OpsNEG{}, O.OpsTRUNC{}, O.OpsADD{}, O.OpsMUL{}, O.OpsSHL{}, O.OpsSHR{}, O.OpsCDIV{}, O.OpsMAX{}, O.OpsCMOD{}, O.OpsCMPLT{}, O.OpsCMPNE{}, O.OpsCMPEQ{}, O.OpsXOR{}, O.OpsOR{}, O.OpsAND{}, O.OpsTHREEFRY{}, O.OpsSUB{}, O.OpsFDIV{}, O.OpsPOW{}, O.OpsFLOORDIV{}, O.OpsFLOORMOD{}, O.OpsWHERE{}, O.OpsMULACC{}], Nil{}},  # 2 ADD AND CDIV CMOD CMPEQ CMPLT CMPNE EXP2 FDIV FLOORDIV FLOORMOD LOG2 MAX MUL MULACC NEG OR POW RECIPROCAL SHL SHR SIN SQRT SUB THREEFRY TRUNC WHERE XOR rej=- slen=0 rlen=0 any=0 nalts=0 name=alu
+    O.PMEntry{2, [O.OpsEXP2{}, O.OpsLOG2{}, O.OpsSIN{}, O.OpsSQRT{}, O.OpsRECIPROCAL{}, O.OpsNEG{}, O.OpsTRUNC{}, O.OpsADD{}, O.OpsMUL{}, O.OpsSHL{}, O.OpsSHR{}, O.OpsCDIV{}, O.OpsMAX{}, O.OpsCMOD{}, O.OpsCMPLT{}, O.OpsCMPNE{}, O.OpsCMPEQ{}, O.OpsXOR{}, O.OpsOR{}, O.OpsAND{}, O.OpsTHREEFRY{}, O.OpsSUB{}, O.OpsFDIV{}, O.OpsPOW{}, O.OpsFLOORDIV{}, O.OpsFLOORMOD{}, O.OpsWHERE{}, O.OpsMULACC{}, O.OpsCONST{}], Nil{}},  # 2 ADD AND CDIV CMOD CMPEQ CMPLT CMPNE EXP2 FDIV FLOORDIV FLOORMOD LOG2 MAX MUL MULACC NEG OR POW RECIPROCAL SHL SHR SIN SQRT SUB THREEFRY TRUNC WHERE XOR rej=- slen=0 rlen=0 any=0 nalts=0 name=alu
   ]}
 
 
@@ -320,7 +320,7 @@
   O.PMEntrys{[
     O.PMEntry{0, [O.OpsINDEX{}], Nil{}},  # 0 INDEX rej=- slen=0 rlen=1 any=0 nalts=1 name=idx
     O.PMEntry{1, [O.OpsAFTER{}], Nil{}},  # 1 AFTER rej=- slen=0 rlen=1 any=0 nalts=1 name=a
-    O.PMEntry{2, [O.OpsEND{}], Nil{}},  # 2 END rej=- slen=0 rlen=1 any=0 nalts=1 name=a
+    O.PMEntry{2, [O.OpsSTORE{}], Nil{}},  # 2 END rej=- slen=0 rlen=1 any=0 nalts=1 name=a
   ]}
 
 
@@ -350,7 +350,7 @@
     O.PMEntry{7, [O.OpsINDEX{}], [O.OpsINDEX{}]},  # 7 INDEX rej=INDEX slen=0 rlen=1 any=0 nalts=1 name=idx2
     O.PMEntry{8, [O.OpsINDEX{}], [O.OpsINDEX{}]},  # 8 INDEX rej=INDEX slen=0 rlen=1 any=0 nalts=1 name=idx2
     O.PMEntry{9, [O.OpsALLREDUCE{}], Nil{}},  # 9 ALLREDUCE rej=- slen=1 rlen=1 any=0 nalts=1 name=red
-    O.PMEntry{10, [O.OpsREDUCE{}], Nil{}},  # 10 REDUCE rej=- slen=1 rlen=1 any=0 nalts=1 name=reduce
+    O.PMEntry{10, [O.OpsWHERE{}], Nil{}},  # 10 REDUCE rej=- slen=1 rlen=1 any=0 nalts=1 name=reduce
     O.PMEntry{11, [O.OpsDETACH{}, O.OpsCONTIGUOUS_BACKWARD{}], Nil{}},  # 11 CONTIGUOUS_BACKWARD DETACH rej=- slen=0 rlen=0 any=0 nalts=0 name=x
     O.PMEntry{12, [O.OpsSINK{}], Nil{}},  # 12 SINK rej=- slen=0 rlen=0 any=0 nalts=0 name=x
     O.PMEntry{13, [O.OpsCOPY{}], Nil{}},  # 13 COPY rej=- slen=0 rlen=1 any=0 nalts=1 name=copy
